x-pack/plugins/enterprise_search/public/applications/app_search/components/schema/schema_meta_engine_logic.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { kea, MakeLogicType } from 'kea';
+
+import { SchemaBaseLogic, SchemaBaseValues, SchemaBaseActions } from './schema_base_logic';
+import { MetaEngineSchemaApiResponse } from './types';
+
+interface MetaEngineSchemaValues extends SchemaBaseValues {
+  fields: MetaEngineSchemaApiResponse['fields'];
+  conflictingFields: MetaEngineSchemaApiResponse['conflictingFields'];
+  conflictingFieldsCount: number;
+  hasConflicts: boolean;
+}
+
+interface MetaEngineSchemaActions extends SchemaBaseActions {
+  loadMetaEngineSchema(): void;
+  onMetaEngineSchemaLoad(response: MetaEngineSchemaApiResponse): MetaEngineSchemaApiResponse;
+}
+
+export const MetaEngineSchemaLogic = kea<
+  MakeLogicType<MetaEngineSchemaValues, MetaEngineSchemaActions>
+>({
+  path: ['enterprise_search', 'app_search', 'meta_engine_schema_logic'],
+  connect: {
+    values: [SchemaBaseLogic, ['dataLoading', 'schema']],
+    actions: [SchemaBaseLogic, ['loadSchema', 'setSchema']],
+  },
+  actions: {
+    loadMetaEngineSchema: true,
+    onMetaEngineSchemaLoad: (response) => response,
+  },
+  reducers: {
+    fields: [
+      {},
+      {
+        onMetaEngineSchemaLoad: (_, { fields }) => fields,
+      },
+    ],
+    conflictingFields: [
+      {},
+      {
+        onMetaEngineSchemaLoad: (_, { conflictingFields }) => conflictingFields,
+      },
+    ],
+  },
+  selectors: {
+    conflictingFieldsCount: [
+      (selectors) => [selectors.conflictingFields],
+      (conflictingFields) => Object.keys(conflictingFields).length,
+    ],
+    hasConflicts: [
+      (selectors) => [selectors.conflictingFieldsCount],
+      (conflictingFieldsCount) => conflictingFieldsCount > 0,
+    ],
+  },
+  listeners: ({ actions }) => ({
+    loadMetaEngineSchema: async () => {
```

Comment:
No, because `loadSchema` isn't setting `fields` or `conflictingFields` (nor does it even have the concept of those two values). Same comment as https://github.com/elastic/kibana/pull/99548#discussion_r629485168.

FWIW, this would be exactly the same architecture if we were using the `.extend()` API. The architectural problem I'm solving here is that these two separate files call the **same** API endpoint and get two bizarrely different set of responses based on the engine type. They contain almost 2 completely separate sets of data except for the 1 shared `schema` value. As such, the `schema` value and the API fetch is shared, but nothing else is (or should be, IMO) to make the code extremely clean and keep indexed vs meta engine logic separate.

Reply:
Jason has clarified this for me in a Zoom call and I have seen the light! ✨